### PR TITLE
[CodeGen] Enable fusion with consumer elementwise ops

### PR DESCRIPTION
**Pull request by antiagainst**
This commit adjusts dispatch region formation logic to
allow fusing operands of root ops. Given linalg.generic
as a root and consumer op, this enables fusing producer
convolution ops into the same region.

This fixes the MobileNet v2 performance regression on
the tensor path: the number of kernel dispatches for
MobileNet v2 goes from 146 down to 103; the latency
goes from 30+ms down to 18ms. (On the buffer path we
were at ~20ms.)

### CODE REVIEW

File: iree/compiler/Conversion/Common/LinalgBufferizePass.cpp [hunk 1]
```diff
@@ -197,7 +199,19 @@ static LogicalResult convertAnyLinalgOp(
...
                                        newOutputBuffers, bvm))) {
-    assert(false);
+    LLVM_DEBUG(llvm::dbgs()
+               << "failed to allocate output buffers for op: " << op << "\n");
+    return failure();
+  }
+
+  // If this is a fill op that acts as a tensor initializer to other linalg ops,
+  // check whether the initialization is used. If not we can just drop the op.
+  if (auto fillOp = dyn_cast<linalg::FillOp>(op.getOperation())) {
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/Common/LinalgBufferizePass.cpp:209]: This seems like the wrong place to do this. Its effectively a dead op? This should move upstream. At least not during bufferization. Its already fairly complex.

File: iree/compiler/Conversion/Common/LinalgBufferizePass.cpp [hunk 2]
```diff
@@ -540,21 +554,42 @@ LogicalResult preProcessInterfaceStoreTensorOp(
...
+
+    // If the result is mapped to a buffer, the corresponding output tensor can
+    // be mapped to the same buffer to make this an inplace update.
     if (resultBuffer && outTensor.hasOneUse()) {
       bvm.map(outTensor, resultBuffer);
     }
+
+    // If the output tensor is not actually used (for initialization) by this
+    // op, we can reuse the result tensor's buffer for some operand.
+    if (!op.payloadUsesValueFromOutputOperandIndex(resultIndex)) {
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/Common/LinalgBufferizePass.cpp:578]: Can you give me an example of why this is needed. This is not the intended contract of `out` tensor. Think of TiedOpInterface. The `out` operand is tied to the result. Tie-ing it to one of the operands is strange.
Some use case here would help, cause I cant think of a reason to have this if you are only fusing named ops with consumers.

File: iree/compiler/Conversion/LinalgToSPIRV/ConcretizeTileAmongWorkgroupsPass.cpp [hunk 3]
```diff
@@ -463,6 +463,12 @@ class ConcretizeTileAmongWorkgroupsPass
       llvm::dbgs() << "\n";
     });
 
+    // Make sure we don't have more than 3 dimensions.
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/ConcretizeTileAmongWorkgroupsPass.cpp:466]: It would be better to rebase this on https://github.com/google/iree/pull/5283. You might not need this after that.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 4]
```diff
@@ -879,24 +932,35 @@ struct MakeDispatchWorkgroupsOp : public RewritePattern {
...
+    auto map = consumer.getIndexingMap(consumerOperand.getOperandNumber());
+    if (!map.isIdentity()) return false;
+
+    // If the producer's result is used by the consumer as an input, fuse if
+    // this producer has no other users.
+    return consumerOperand.get().hasOneUse();
+  } else {
+    // If the producer's result is used by the consumer as an output
+    // initializer, fuse if the producer is a parallel operation.
+    return isElementWiseParallelOp(producer);
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:962]: The comment is fairly disjoint w.r.t what I read as the the implementation of `isElementWiseParallelOp`

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 5]
```diff
@@ -144,6 +144,34 @@ static SmallVector<Value, 4> convertToWorkload(OpBuilder &b, Location loc,
   return workload;
 }
 
+/// Returns the fusion groups for the given `op`.
+static SmallVector<int64_t, 1> getFunsionGroups(Operation *op) {
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:148]: Typo : `getFunsionGroups` -> `getFusionGroups`

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Awesome! Overall looks great. Few comments to resolve!

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp (hunk 4 above)
Comment (antiagainst) [reply iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:962]: Moved `isElementWiseParallelOp` down here and modified the comment a bit.

File: iree/compiler/Conversion/Common/LinalgBufferizePass.cpp (hunk 1 above)
Comment (antiagainst) [reply iree/compiler/Conversion/Common/LinalgBufferizePass.cpp:209]: Hmm, looks like this is not needed now; maybe because some recent landed changes since I've this implemented. I'll remove this.

But good point. I'll still send this out as a general canonicalization pattern upstream in Linalg.

**Review by hanhanW — APPROVED** [review]

File: iree/compiler/Conversion/Common/LinalgBufferizePass.cpp (hunk 2 above)
Comment (antiagainst) [reply iree/compiler/Conversion/Common/LinalgBufferizePass.cpp:578]: It's for eliding intermediate buffers needed between `linalg.conv` and its consumer `linalg.generic`. Without this, for the following fused IR in the same dispatch region:

```mlir
scf.for
  scf.for
    scf.for
      %init0 = linalg.init_tensor
      %init1 = linalg.init_tensor
      %fill = linalg.fill(%init1, ...)
      %conv = linalg.conv_2d_input_nhwc_filter_hwcf outs(%fill)
      %generic = linalg.generic ins(%conv, ...) outs(%init0)
```

After bufferization, we'll see an allocation for the result for `%conv`:


```mlir
scf.for
  scf.for
    scf.for
      %subview = memref.subview
      %alloc = memref.alloc
      linalg.fill(%alloc, ...)
      linalg.conv_2d_input_nhwc_filter_hwcf outs(%alloc)
      %generic = linalg.generic ins(%alloc, ...) outs(%subview)
```

The allocation is breaking the lowering to GPU and it's also unnecessary as we can just directly use `%subview`. The connection here is to basically tell the producer `%conv` to reuse the buffer for `%subview`.

I'm also feeling a bit uneasy about this.. But this seems the best way to make the connection, without performing more complicated analysis afterwards to elide the allocation. If you think there are better ways to handle this, I'm certainly happy to try out.

File: iree/compiler/Conversion/LinalgToSPIRV/ConcretizeTileAmongWorkgroupsPass.cpp (hunk 3 above)
Comment (antiagainst) [reply iree/compiler/Conversion/LinalgToSPIRV/ConcretizeTileAmongWorkgroupsPass.cpp:466]: For now will leave this as-is. Would like to land this soon if possible so we can have a checkpoint for evaluating performance impacts of upcoming refactorings.

File: iree/compiler/Conversion/Common/LinalgBufferizePass.cpp (hunk 2 above)
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/Common/LinalgBufferizePass.cpp:568]: Yeah, I suspected this was the case. This can get tricky though. Best approach to try here is to mark these as TiedOperands at dispatch region creation (and use that information here).
The advantage there would be if we can tie the operands, we just bail on the fusion. So we wont hit a case where at a later stage we have to allocate a buffer.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp (hunk 5 above)
Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp]: THis doesnt seem to be resolved.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 6]
```diff
@@ -317,10 +344,36 @@ static void pullInProducersInSameGroup(
...
           maybeFusionInfo->fusedProducer.getOperation()->removeAttr(
               kFusionGroupsAttr);
+          fusedProducer = maybeFusionInfo->fusedProducer;
+        }
+      }
+
+      // If the producer is successfully fused, go recursive over the current
+      // producer's operands and pull them in if they are marked to be fused
+      // into the current group.
+      if (fusedProducer) {
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:360]: Maybe leave a TODO here. THis is strange layering coming from the fact that this method is trying to see if it can pull in all its producers. It should actually just be trying to pull in one producer at a time. Then we dont need these two `if` conditions.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 7]
```diff
@@ -917,26 +981,24 @@ static void decideFusableLinalgOps(FuncOp funcOp) {
...
+      // In that case, it has fusion groups attached on it.
+      SmallVector<int64_t, 1> fusionGroups = getFunsionGroups(op);
+      if (fusionGroups.empty()) {
+        // Otherwise, create a new group.
+        unsigned newGroup = numRootOps++;
+        fusionGroups.push_back(newGroup);
+        op->setAttr(kRootOpAttr, builder.getI64IntegerAttr(newGroup));
+      }
+
+      for (OpOperand &operand : linalgOp.getShapedOpOperands()) {
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:994]: So we need to make a distinction between case where elementwise operation is the root operation. We should only be pulling in the producer if the producer has only one use. If the producer has multiple uses its better to leave the producer in its own dispatch region. Think of a matmul which is used in two elementwise operations. You dont want to fuse it with both (the computation is replicated).

Comment (MaheshRavishankar) [reply iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:994]: Also add a test for this kind of pattern.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp [hunk 8]
```diff
@@ -917,26 +981,24 @@ static void decideFusableLinalgOps(FuncOp funcOp) {
       // Start with a root operation and fuse its producers.
       Operation *op = linalgOp.getOperation();
       if (!isRootOp(op)) continue;
-      if (op->getAttrOfType<ArrayAttr>(kFusionGroupsAttr)) continue;
-      unsigned currGroupNum = numRootOps++;
-      op->setAttr(kRootOpAttr, builder.getI64IntegerAttr(currGroupNum));
-      for (auto operand : linalgOp.getShapedOperands()) {
-        auto producer = operand.getDefiningOp<linalg::LinalgOp>();
+
+      // This might be a root op that is already fused into another root op.
```
Comment (MaheshRavishankar) [inline iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:984]: I think it might be better to add a flag that guards when we chose to fuse elementwise operations with the named ops. There is a bit of an issue with aggressive fusion (number of descriptors used, etc. etc.). We can for now leave the flag on by default. Eventually we need to have some arbitrary heuristic to limit fusion (number of operands of dispatch region, for example), but that is not needed right here.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp (hunk 5 above)
Comment (antiagainst) [reply iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp]: It's done in ba4ec3c.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp (hunk 6 above)
Comment (antiagainst) [reply iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:370]: Added TODO. Also removed the `if` condidtions here: we have the check in the recursive call anyway. Having it here avoids ever going into recursion if not necessary; but the gain is marginal and it complicates the logic.

File: iree/compiler/Conversion/Common/LinalgBufferizePass.cpp (hunk 2 above)
Comment (antiagainst) [reply iree/compiler/Conversion/Common/LinalgBufferizePass.cpp:568]: I'm not sure how we can mark them as tied in dispatch region formation, given they are *internal* tensors to the `flow.dispatch.workgroups` op. The result of `%conv` is only used by the consumer `%generic`. It does not escape to the `flow.dispatch.workgroups` interface so there is no block argument for it.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp (hunk 7 above)
Comment (antiagainst) [reply iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:993]: Yup. We only fuse the producer when it just has one user. This is done via `isProducerFusable`, which has the check `hasOneUse`. There is already a test for it, `dont_fuse_conv2d_with_multiple_uses` in `disaptch_linalg_on_tensors.mlir`.

File: iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp (hunk 8 above)
Comment (antiagainst) [reply iree/compiler/Dialect/Flow/Transforms/DispatchLinalgOnTensors.cpp:983]: Added a CL option to disable operand fusion.

**Comment by MaheshRavishankar** [comment]
Apart from the tied operands stuff, everything looks fine here. DO you have a solution for that?

**Comment by antiagainst** [comment]
> Apart from the tied operands stuff, everything looks fine here. DO you have a solution for that?

I have replied previously: https://github.com/google/iree/pull/5282#discussion_r606354086. I don't think it's possible to do it at dispatch region formation time. The other way I can think of is to do it afterwards with a pattern to look at buffers and try to elide them if they are of such usage. It will keep the bufferization clean, but at the cost of more code. Not sure it's better.

**Comment by MaheshRavishankar** [comment]
> > Apart from the tied operands stuff, everything looks fine here. DO you have a solution for that?
> 
> I have replied previously: [#5282 (comment)](https://github.com/google/iree/pull/5282#discussion_r606354086). I don't think it's possible to do it at dispatch region formation time. The other way I can think of is to do it afterwards with a pattern to look at buffers and try to elide them if they are of such usage. It will keep the bufferization clean, but at the cost of more code. Not sure it's better.

ok, lets leave it here. Might get very tricky, but lets see how it works out.

**Review by MaheshRavishankar — APPROVED** [review]
Looks good to me. Lets keep this turned off by default since there are issues with descriptors, etc. which might just make things less stable.

**Comment by antiagainst** [comment]
> ok, lets leave it here. Might get very tricky, but lets see how it works out.

+1!

> Looks good to me. Lets keep this turned off by default since there are issues with descriptors, etc. which might just make things less stable.

Further restricted the fusion to only fuse one input in 0389bb0 to be less aggressive. This will help. I'd like to leave it on though given all tests are passing. Default off just means it won't really be tested extensively and there are more chances to regress. We can turn it to default off if this really turns out to be problematic.

**Comment by hanhanW** [comment]
hmm, I think this breaks MobileNetV2 benchmark in Linalg on tensors.

https://buildkite.com/iree/iree-android-benchmark/builds/2224#f0c20f89-02d6-4e63-bd39-e33b16b67867

[Full log](https://buildkite.com/organizations/iree/pipelines/iree-android-benchmark/builds/2224/jobs/f0c20f89-02d6-4e63-bd39-e33b16b67867/raw_log) shows that `error: failed to legalize operation 'vector.transfer_read'`

I thought this is because we didn't add `-iree-spirv-enable-vectorization`, and somehow a vector op was created, see #5190. However, I added the  flag and got compile error too. And it works if I remove `--iree-vulkan-target-triple=valhall-g77-unknown-android10`. Could you check if this works on your local side?

To repro:

```
iree-translate --iree-mlir-to-vm-bytecode-module --iree-hal-target-backends=vulkan-spirv ~/mobilenet-v2/iree_input.mlir -o a.vmfb -iree-flow-dispatch-linalg-on-tensors -iree-codegen-spirv-experimental-linalg-on-tensors -iree-spirv-enable-vectorization --iree-vulkan-target-triple=valhall-g77-unknown-android10
```

**Comment by antiagainst** [comment]
The reason is constant inlining. We can inline several large constants in the same kernel right now. After bufferization we'll have `memref.buffer_cast` for them. Those `memref.buffer_cast` ops aren't recognized by `VectorizeMemRefPass`, which expects `hal.interface.binding.subspan` ops and turns `vector.transfer_read` into normal `load` ops.

The interesting thing is that I wasn't seeing this before with compiling the full model and benchmarking it. It turns out that the model I was using actually has fake weights. I generated by following:

https://github.com/google/iree/blob/ccae793f8c7c208a0b1b0baf489cfd121e5f7b4c/integrations/tensorflow/e2e/keras/applications/BUILD#L33-L57

I previously thought that'd use the real weights given the `imagenet` parameter. But actually it is different from 

https://github.com/google/iree/blob/ccae793f8c7c208a0b1b0baf489cfd121e5f7b4c/integrations/tensorflow/e2e/keras/applications/applications_test.py#L30-L40

There is a `use_external_weights` option that need to be turned on to use real weights..

**Comment by antiagainst** [comment]
It might take some time before I come up with a proper fix. In the meanwhile we can use `-iree-flow-inline-constants-max-byte-length=16` to unbreak.
